algebra/vector_space.lean

```diff
@@ -0,0 +1,305 @@
+theorem zero_add : 0 + A = A := ext (λ v, by simp [zero_add])
+theorem add_right_neg : A + (-A) = 0 := ext (λ v, by simp [add_right_neg])
+theorem add_left_neg : (-A) + A = 0 := ext (λ v, by simp [add_left_neg])
+theorem add_assoc : A + B + C = A + (B + C) := ext (λ v, by simp [add_assoc])
+theorem add_comm : A + B = B + A := ext (λ v, by simp [add_comm])
+theorem smul_left_distrib : c⬝(A+B) = c⬝A + c⬝B := ext (λ v, by simp [smul_left_distrib])
+theorem smul_right_distrib : (c+d)⬝A = c⬝A + d⬝A := ext (λ v, by simp [smul_right_distrib])
+theorem mul_smul : (c*d)⬝A = c⬝(d⬝A) := ext (λ v, by simp [mul_smul])
+theorem one_smul : 1⬝A = A := ext (λ v, by simp [one_smul])
+
+instance : vector_space K (linear_space K V W) :=
+{
+    add                 := add,
+    add_assoc           := add_assoc,
+    zero                := zero,
+    zero_add            := zero_add,
+    add_zero            := add_zero,
+    neg                 := neg,
+    add_left_neg        := add_left_neg,
+    add_comm            := add_comm,
+    smul                := smul,
+    smul_left_distrib   := smul_left_distrib,
+    smul_right_distrib  := smul_right_distrib,
+    mul_smul            := mul_smul,
+    one_smul            := one_smul
+}
+
+end Hom
+
+def Hom {K : Type} (V W : Type) [field K] [vector_space K V] [vector_space K W] :=
+vector_space K (linear_space K V W)
+
+def vector_space.dual {K: Type} (V: Type) [field K] [vector_space K V] :=
+@Hom K V K _ _ field.to_vector_space
+
+-- \ast
+postfix `∗`:100 := vector_space.dual
+
+
+section matrix_ring
+
+variables {v : V} {A B C : linear_space K V V}
+
+def id : linear_space K V V :=
+{
+    T       := id,
+    map_add := (λ u v, rfl),
+    map_mul := (λ u v, rfl),
+}
+
+def comp (A B : linear_space K V V) : linear_space K V V :=
+{
+    T       := (λ v, A.T $ B.T v),
+    map_add := (λ u v, by simp),
+    map_mul := (λ c v, by simp)
+}
+
+instance : has_mul (linear_space K V V) := ⟨comp⟩
+
+@[simp] lemma id_simp : (id : linear_space K V V).T v = v := rfl
+@[simp] lemma comp_simp : (A * B).T v = A.T (B.T v) := rfl
+
+variables (v A B C)
+
+@[simp] lemma comp_id : A * id = A := ext (λ v, rfl)
+@[simp] lemma id_comp : id * A = A := ext (λ v, rfl)
+theorem comp_assoc : (A * B) * C = A * (B * C) := ext (λ v, rfl)
+theorem comp_add : A * (B + C) = A * B + A * C := ext (λ v, by simp)
+theorem add_comp : (A + B) * C = A * C + B * C := ext (λ v, rfl)
+
+instance : ring (linear_space K V V) :=
+{
+    add             := add,
+    add_assoc       := add_assoc,
+    zero            := zero,
+    zero_add        := zero_add,
+    add_zero        := add_zero,
+    neg             := neg,
+    add_left_neg    := add_left_neg,
+    add_comm        := add_comm,
+    mul             := comp,
+    mul_assoc       := comp_assoc,
+    one             := id,
+    one_mul         := id_comp,
+    mul_one         := comp_id,
+    left_distrib    := comp_add,
+    right_distrib   := add_comp
+}
+
```

Comment:
In the future we should handle most of linear algebra more generally and use the `module` structure, then we can add `K-algebras`.